cost_rules/app.py

```diff
@@ -0,0 +1,263 @@
+import json
+import logging
+import os
+import re
+
+import boto3
+import requests
+
+
+LOG = logging.getLogger(__name__)
+LOG.setLevel(logging.INFO)
+
+# This is global so that it can be stubbed in test.
+# Because this is global its value will be retained
+# in the lambda environment and re-used on warm runs.
+org_client = None
+
+
+def _get_os_var(varnam):
+    try:
+        return os.environ[varnam]
+    except KeyError as exc:
+        raise Exception(f"The environment variable '{varnam}' must be set")
+
+
+def _parse_env_list(string):
+    '''
+    Unpack a CSV into a list of strings.
+
+    In order to pass a list of strings through an environment variable,
+    it needs to be encoded in a single string, use CSV.
+    '''
+    return string.split(',')
+
+
+def _strip_special_chars(value):
+    '''
+    The name of a cost category must adhere to: ^(?! )[\p{L}\p{N}\p{Z}-_]*(?<! )$
+
+    Replace any disallowed characters with '_'
+    '''
+    return re.sub('[^a-zA-Z0-9 -]', '_', value)
+
+
+def collect_account_tag_codes(tag_names):
+    '''
+    Query account tags for fallback values if the resource is untagged.
+    '''
+
+    account_codes = {}
+
+    # create boto client
+    global org_client
+    if org_client is None:
+        org_client = boto3.client('organizations')
+
+    # get list of accounts
+    account_pages = org_client.get_paginator('list_accounts').paginate()
+
+    # check for tags on each account
+    for account_page in account_pages:
+        for account in account_page['Accounts']:
+            account_id = account['Id']
+
+            tag_pager = org_client.get_paginator('list_tags_for_resource')
+            tag_pages = tag_pager.paginate(ResourceId=account_id)
+
+            found = None
+            for tag_page in tag_pages:
+                for tag in tag_page['Tags']:
+                    if tag['Key'] in tag_names:
+                        # get a 6-digit numeric code from the tag
+                        found = re.search(r'[0-9]{6}', tag['Value'])
+
+                        if found is None:
+                            LOG.warning(f"No numeric code found in tag: {tag['Value']}")
```

Comment:
this is only looking at `CostCenter` or `CostCenterOther` tags because of the `if` on line 71 above, so I think it's appropriate to log this as a warning if those tags exist but don't have a valid code in them

Reply:
i see. looks good then.